drivers/clk/clk-altera-c5-fpll.c

```
@@ -0,0 +1,290 @@
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	m_int = m >> fpll->fractional_carry_bit;
+
+	reg = m_int / 2;
+	reg |= ((m_int / 2) + (m_int % 2)) << 8;
+
+	if (m_int % 2)
+		reg |= (1 << 17);
+
+	writel(0x01, fpll->base + C5_FPLL_MODE_REG_ADDR);
+	writel(0x10101, fpll->base + C5_FPLL_N_CNTR_ADDR);
+	writel(reg, fpll->base + C5_FPLL_M_CNTR_ADDR);
+	writel(m_frac, fpll->base + C5_FPLL_M_CNTR_FRACTIONAL_VAL_ADDR);
+
+	chan->rate = rate;
+
+	for (i = 0; i < fpll->num_channels; i++) {
+		c = DIV_ROUND_CLOSEST_ULL(vco, fpll->channels[i].rate);
+		fpll->channels[i].rate = DIV_ROUND_CLOSEST_ULL(vco, c);
+
+		reg = c / 2;
+		reg |= ((c / 2) + (c % 2)) << 8;
+		reg |= C5_FPLL_REG_C_CNTR(fpll->channels[i].num);
+
+		if (c % 2)
+			reg |= BIT(17);
+
+		writel(reg, fpll->base + C5_FPLL_C_CNTR_ADDR);
+	}
+
+	writel(0x00u, fpll->base + C5_FPLL_START_REG_ADDR);
+
+	ret = readl_poll_timeout(fpll->base + C5_FPLL_STATUS_REG_ADDR,
+				 status, status & 0x00000001, 1000, 30000);
+	if (ret < 0)
+		return ret;
+
+	return 0;
+}
+
+static unsigned long altera_c5_fpll_recalc_rate(struct clk_hw *hw,
+						unsigned long parent_rate)
+{
+	struct altera_c5_fpll_chan *chan = to_channel(hw);
+
+	return chan->rate;
+}
+
+static const struct clk_ops altera_c5_fpll_ops = {
+	.recalc_rate = altera_c5_fpll_recalc_rate,
+	.round_rate = altera_c5_fpll_round_rate,
+	.set_rate = altera_c5_fpll_set_rate,
+};
+
+static int altera_c5_fpll_parse_dt(struct altera_c5_fpll *fpll)
+{
+	struct device_node *np = fpll->dev->of_node, *chan_np;
+	unsigned int cnt = 0;
+	u32 tmp;
+	int ret;
+
+	fpll->ref_clk = devm_clk_get(fpll->dev, NULL);
+	if (IS_ERR(fpll->ref_clk))
+		return PTR_ERR(fpll->ref_clk);
+
+	fpll->num_channels = of_get_available_child_count(np);
+	if (fpll->num_channels > C5_FPLL_NUM_CHAN)
+		return -EINVAL;
+
+	for_each_available_child_of_node(np, chan_np) {
+		of_property_read_u32(chan_np, "reg",
+				     &fpll->channels[cnt].num);
+		of_property_read_u32(chan_np, "adi,rate",
```

We should validate `reg` and `adi,rate`. Furthermore, they are mandatory properties which means we should error out if they are not provided.